src/Mod/Sketcher/Gui/DrawSketchHandlerScale.h

```diff
@@ -0,0 +1,529 @@
+            // Add the lines to show lengths
+            addLineToShapeGeometry(toVector3d(referencePoint), toVector3d(startPoint), true);
+
+            addLineToShapeGeometry(toVector3d(referencePoint), toVector3d(endPoint), true);
+        }
+        else {
+            int firstCurveCreated = getHighestCurveIndex() + 1;
+
+            const std::vector<Constraint*>& vals = Obj->Constraints.getValues();
+            // avoid applying equal several times if cloning distanceX and distanceY of the
+            // same part.
+            std::vector<int> geoIdsWhoAlreadyHasEqual = {};
+
+            for (auto& cstr : vals) {
+                int firstIndex = indexInVec(listOfGeoIds, cstr->First);
+                int secondIndex = indexInVec(listOfGeoIds, cstr->Second);
+                int thirdIndex = indexInVec(listOfGeoIds, cstr->Third);
+
+                auto newConstr = std::unique_ptr<Constraint>(cstr->copy());
+                newConstr->First = firstCurveCreated + firstIndex;
+
+                if ((cstr->Type == Symmetric || cstr->Type == Tangent || cstr->Type == Perpendicular
+                     || cstr->Type == Angle)
+                    && firstIndex >= 0 && secondIndex >= 0 && thirdIndex >= 0) {
+                    newConstr->Second = firstCurveCreated + secondIndex;
+                    newConstr->Third = firstCurveCreated + thirdIndex;
+                }
+                else if ((cstr->Type == Coincident || cstr->Type == Tangent
+                          || cstr->Type == Symmetric || cstr->Type == Perpendicular
+                          || cstr->Type == Parallel || cstr->Type == Equal || cstr->Type == Angle
+                          || cstr->Type == PointOnObject || cstr->Type == InternalAlignment)
+                         && firstIndex >= 0 && secondIndex >= 0
+                         && thirdIndex == GeoEnum::GeoUndef) {
+                    newConstr->Second = firstCurveCreated + secondIndex;
+                }
+                else if ((cstr->Type == Radius || cstr->Type == Diameter || cstr->Type == Weight)
+                         && firstIndex >= 0) {
+                    newConstr->setValue(newConstr->getValue() * scaleFactor);
+                }
+                else if ((cstr->Type == Distance || cstr->Type == DistanceX
+                          || cstr->Type == DistanceY)
+                         && firstIndex >= 0 && secondIndex >= 0) {
+                    newConstr->Second = firstCurveCreated + secondIndex;
+                    newConstr->setValue(newConstr->getValue() * scaleFactor);
+                }
+                else if ((cstr->Type == Block) && firstIndex >= 0) {
+                    newConstr->First = firstCurveCreated + firstIndex;
+                }
+                else {
+                    continue;
+                }
+
+                ShapeConstraints.push_back(std::move(newConstr));
+            }
+        }
+    }
+
+    int indexInVec(const std::vector<int>& vec, int elem) const

```

Comment:
This kind of function should be moved out to tools for code reuse, as it is not specific of this DSH. Others DSH should use it (I think I remember Rotate also uses it).

Then, I would say it is something more than a generic "indexInVec" as it is for a std::vector<int> which has inside GeoIds. The name of the function should reflect this "IndexOfGeoId" (for example). Then the parameter "element" should perhaps be "geoid".

That would also hint why this function is being written at all (because it handles special values differently), as just getting the index of a vector is a common operation, which is generally implemented via std::find and std::distance.

Reply:
Sounds good, but where should I put this function so that it is shared? In Utils.cpp?

Reply:
That is what I would do.

If you take encapsulation into account, you do not need to access class members, which tells us it should not be a member function of the class or a static member function, so it should be a "free" function outside the class. If we want code reuse it needs to be declared in a header you can include in both places. The header of choice for this kind of functions is now Utils.h.